### DISCUSSION TITLE
Why can't a stackalloc expression be assigned to a pointer field?

### DISCUSSION THREAD

**Discussion by jnm2** [dotnet/csharplang#3894] [General] [answered]
This works:

```cs
char* temp = stackalloc char[1024];

var assemblyInfo = new ASSEMBLY_INFO
{
    pszCurrentAssemblyPathBuf = temp,
};
```

So why not this?
```cs
var assemblyInfo = new ASSEMBLY_INFO
{
    // ❌ CS8346 Conversion of a stackalloc expression of type 'char' to type 'char*' is not possible.
    //                          ↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓
    pszCurrentAssemblyPathBuf = stackalloc char[1024],
};
```

`ASSEMBLY_INFO.pszCurrentAssemblyPathBuf`  is declared as `char*`.

**Comment by 333fred** [comment, marked as answer]
This is because `stackalloc` in C# behaves differently when it's a top-level expression vs a nested expression. Previous to C# 7.2, `stackalloc` was only allowed as top-level expression, and could not be nested. You can see remnants of this here: https://github.com/dotnet/csharplang/blob/master/spec/unsafe-code.md#stack-allocation. For legacy reasons, when you use `stackalloc` as a variable initializer today, you get a `char*` back as the natural type, and we add an implicit conversion to `Span<T>` if the thing you assign to is a `Span<T>`. However, for any other context, we do the safer thing and `stackalloc` always returns a `Span<T>`, instead of a `T*`. This means that in your nested object initializer, you're trying to assign a `Span<char>` to a `char*`, which is of course not valid.
